File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/internal/callback/LogCallbackState.java
package org.maplibre.nativeffi.internal.callback;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.util.Objects;
import org.maplibre.nativeffi.internal.c.MapLibreNativeC;
import org.maplibre.nativeffi.internal.c.mln_log_callback;
import org.maplibre.nativeffi.internal.convert.NativeValues;
import org.maplibre.nativeffi.internal.loader.NativeAccess;
import org.maplibre.nativeffi.internal.memory.MemoryUtil;
import org.maplibre.nativeffi.internal.status.Status;
import org.maplibre.nativeffi.log.LogCallback;
import org.maplibre.nativeffi.log.LogRecord;

/** Owns process-global logging callback state. */
public final class LogCallbackState implements AutoCloseable {
  private static final Object LOCK = new Object();
  private static LogCallbackState current;

  private final Arena arena;
  private final CallbackLifecycle lifecycle = new CallbackLifecycle();
  private final LogCallback callback;
  private final MemorySegment stub;

  private LogCallbackState(LogCallback callback) {
    this.arena = Arena.ofShared();
    this.callback = callback;
    this.stub = mln_log_callback.allocate(this::invoke, arena);
  }

  public static void set(LogCallback callback) {
    NativeAccess.ensureLoaded();
    var replacement = new LogCallbackState(Objects.requireNonNull(callback, "callback"));
    LogCallbackState previous;
    try {
      synchronized (LOCK) {
        if (current != null && current.isCurrentThreadInCallback()) {
          throw Status.callbackReentry("Log callback");
        }
        Status.check(MapLibreNativeC.mln_log_set_callback(replacement.stub, MemorySegment.NULL));
        previous = current;
        current = replacement;
      }
    } catch (RuntimeException | Error error) {
      closeQuietly(replacement);
      throw error;
    }
    closeQuietly(previous);
  }

  public static void clear() {
    NativeAccess.ensureLoaded();
    LogCallbackState previous;
    synchronized (LOCK) {
      if (current != null && current.isCurrentThreadInCallback()) {
        throw Status.callbackReentry("Log callback");
      }
      Status.check(MapLibreNativeC.mln_log_clear_callback());
      previous = current;
      current = null;
    }
    closeQuietly(previous);
  }

  static LogCallbackState currentForTesting() {
    synchronized (LOCK) {
      return current;
    }
  }

  MemorySegment stubForTesting() {
    return stub;
  }

  boolean isCurrentThreadInCallback() {
    return lifecycle.isCurrentThreadInCallback();
  }

  private int invoke(
      MemorySegment userData, int severity, int event, long code, MemorySegment message) {
    var lease = lifecycle.enter();
    if (lease.isEmpty()) {
      return 0;
    }
    try (var ignored = lease.get()) {
      var record =
          new LogRecord(
              NativeValues.logSeverity(severity),
              severity,
              NativeValues.logEvent(event),
              event,
              code,
              MemoryUtil.copyCString(message));
      return callback.log(record) ? 1 : 0;
    } catch (Throwable ignored) {
      return 0;
    }
  }

  @Override
  public void close() {
    lifecycle.close("Log callback", arena::close);
  }

  private static void closeQuietly(LogCallbackState state) {
    if (state == null) {
      return;
    }
    try {
      state.close();
    } catch (IllegalStateException ignored) {
      // Another native thread may be finishing an upcall. The process will reclaim this arena.
    }
  }
}

File: bindings/kotlin/src/nativeMain/kotlin/org/maplibre/nativeffi/internal/callback/LogCallbackState.kt
package org.maplibre.nativeffi.internal.callback

import kotlinx.cinterop.ByteVar
import kotlinx.cinterop.COpaquePointer
import kotlinx.cinterop.CPointer
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.StableRef
import kotlinx.cinterop.asStableRef
import kotlinx.cinterop.staticCFunction
import org.maplibre.nativeffi.internal.c.mln_log_clear_callback
import org.maplibre.nativeffi.internal.c.mln_log_set_callback
import org.maplibre.nativeffi.internal.memory.MemoryUtil
import org.maplibre.nativeffi.internal.status.Status
import org.maplibre.nativeffi.log.LogCallback
import org.maplibre.nativeffi.log.LogEvent
import org.maplibre.nativeffi.log.LogRecord
import org.maplibre.nativeffi.log.LogSeverity

/** Owns process-global logging callback state. */
@OptIn(ExperimentalForeignApi::class)
internal class LogCallbackState(private val callback: LogCallback) : AutoCloseable {
  private val selfRef = StableRef.create(this)
  private val gate = CallbackGate("log callbacks") { selfRef.dispose() }

  fun invoke(severity: UInt, event: UInt, code: Long, message: CPointer<ByteVar>?): UInt {
    val lease = gate.enter() ?: return 0U
    return try {
      val record =
        LogRecord(
          LogSeverity.fromNative(severity),
          LogEvent.fromNative(event),
          code,
          MemoryUtil.copyCString(message),
        )
      if (callback.log(record)) 1U else 0U
    } catch (_: Throwable) {
      0U
    } finally {
      lease.close()
    }
  }

  override fun close() = gate.close()

  internal fun isClosedForTesting(): Boolean = gate.isClosedForTesting()

  internal companion object {
    fun set(callback: LogCallback) {
      val replacement = LogCallbackState(callback)
      try {
        Status.check(
          mln_log_set_callback(
            staticCFunction(::logCallback),
            replacement.selfRef.asCPointer(),
            staticCFunction(::releaseLogCallback),
          )
        )
      } catch (error: Throwable) {
        replacement.close()
        throw error
      }
    }

    fun clear() {
      Status.check(mln_log_clear_callback())
    }
  }
}

@OptIn(ExperimentalForeignApi::class)
private fun logCallback(
  userData: COpaquePointer?,
  severity: UInt,
  event: UInt,
  code: Long,
  message: CPointer<ByteVar>?,
): UInt =
  userData?.asStableRef<LogCallbackState>()?.get()?.invoke(severity, event, code, message) ?: 0U

@OptIn(ExperimentalForeignApi::class)
private fun releaseLogCallback(userData: COpaquePointer?) {
  userData?.asStableRef<LogCallbackState>()?.get()?.close()
}
